File: Go/004/main.go
package main

import (
	"fmt"
	"strconv"
	"time"
)

func palindromeCheck(n int) bool {
	num := strconv.Itoa(n)

	for i := range len(num) / 2 {
		if string(num[i]) != string(num[len(num)-1-i]) {
			return false
		}
	}

	return true
}

func solve() int {
	var answer int

	for i := 100; i < 1000; i++ {
		for j := 100; j < 1000; j++ {
			n := i * j
			if palindromeCheck(n) && n > answer {
				answer = n
			}
		}
	}

	return answer
}

func main() {
	start := time.Now()
	answer := solve()
	elapsed := time.Since(start)

	fmt.Println(answer)
	fmt.Println("--------------------------------------------------")
	fmt.Printf("  %.6f seconds\n", elapsed.Seconds())
}

File: Python/004/main.py
import time


limit = 1000


def palindromeCheck(n):
    pal = str(n)

    return pal == pal[::-1]


def solve(to):
    answer = 0

    for i in range(to - 1, 99, -1):
        if i * i <= answer:
            break

        for j in range(i - 1, 99, -1):
            product = i * j

            if product <= answer:
                break

            if palindromeCheck(i * j):
                answer = i * j

    return answer


# The answer should be 906609
start = time.time()
answer = solve(limit)
end = time.time()
print(answer)
print("--------------------------------------------------")
print(f"  {end - start:.6f} seconds")
